docs/internal/release-versions.md

```diff
@@ -0,0 +1,167 @@
+# Release Versions
+
+* [Introduction](#introduction)
+* [Version format](#version-format)
+* [Build version](#build-version)
+* [Compatibility](#compatibility)
+* [Examples](#examples)
+* [Alternatives considered](#alternatives-considered)
+
+## Introduction
+
+We have several pieces of software:
+
+* Chain Core itself
+* SDKs (currently Java, with Ruby and Node.js in the roadmap)
+* Mac app
+* Windows installer
+* Docker container
+
+Each of these is going to be updated for one of three reasons:
+
+* its own improvements (e.g., a bugfix in the SDK or a Mac app)
+* improvements in a product as a whole (e.g. a new feature in Chain Core with updated SDK support)
+* as a dependency on another package (e.g. fix in a Chain Core needs corresponding Mac and Windows executables)
+
+We need a consistent, yet simple, versioning scheme that allows both individual updates as well as grouped updates: from major feature releases to smaller, maintenance releases.
+
+We probably do not want to formally commit to specific semantics of the version numbers, such as [SemVer](http://semver.org), but we need to keep all numbers in order, so it’s easier navigate them.
+
+## Version format
+
+Use the following version format:
+
+    <major>.<minor>.<build>
+
+**Major** and **minor** versions are reserved for the entire product updates usually affecting several packages. Either version can be bumped based on the significance of the update. Individual packages must not use either **major** or **minor** version on their own without synchronization with the entire product line, but instead should use **build** version.
+
+**Build** version must be formatted as dates in the following format: `YYYYMMDD` or omitted entirely. **Major** and **minor** versions are plain integers.
+
+Examples:
+
+* 1.0
+* 1.0.20161024
+* 1.7.20170119
+* 2.3.20170601
+
+
+## Build version
+
+The reason to use dates for **build** versions is to avoid frequent collisions across different software packages that may convey non-existant correlation. E.g. if at some point Java SDK and Mac app both have version `1.3.2`, some may be confused by the identical version. 
+
+On the other hand, dependent releases usually happen on the same day and therefore most likely will have the identical versions, indicating the correlation truthfully. E.g. if Chain Core adds a minor fix in the API, all SDKs may be updated with the same new build number: `1.0.161205`.
+
+## Compatibility
+
+We should strive to keep expectation that minor version updates do not break backward compatibility. E.g. if we need to introduce a completely new SDK, we should bump the major version, or rename the SDK. If we need to retire/improve all APIs, we should add a compatibility layer and diagnostic messages to the SDK without simply removing them.
```

Comment:
I read this as "if any user code will break with a given change, upgrade the major version".

Taking stock of that rule, plus the breadth of the software (spanning the client API, network API, and all SDKs), it seems as if Chain Core will undergo relatively many major version changes, at least compared to other most other software. Sampling at random, I will note that Postgres is on major version 9 after 20 years of development, and that Ruby on Rails is on version 5 after over a decade. Given that we have lots of ideas for Chain Core's evolution, it wouldn't surprise me if we were on version 50 this time next year.

I think this is _totally fine_, but since it is a little unconventional, it just feels worth pointing out. I sort of assume we will get used to upgrading the major version fairly casually.